tools/unitctl/unit-client-rs/src/unit_client.rs

```diff
@@ -223,72 +211,73 @@ impl UnitClient {
-                        path: path.to_string(),
-                    })?;
-
-                return Err(UnitClientError::HttpResponseJsonBodyError {
-                    status,
+        let response = response_future
+            .await
+            .map_err(|error| UnitClientError::new(error, self.control_socket.to_string(), path.to_string()))?;
+
+        let status = response.status();
+        let body = hyper::body::aggregate(response)
+            .await
+            .map_err(|error| UnitClientError::new(error, self.control_socket.to_string(), path.to_string()))?;
+        let reader = &mut body.reader();
+        if !status.is_success() {
+            let error: HashMap<String, String> =
+                serde_json::from_reader(reader).map_err(|error| UnitClientError::JsonError {
+                    source: error,
                     path: path.to_string(),
-                    error: error.get("error").unwrap_or(&"Unknown error".into()).to_string(),
-                    detail: error.get("detail").unwrap_or(&"".into()).to_string(),
-                });
-            }
-            serde_json::from_reader(reader).map_err(|error| UnitClientError::JsonError {
-                source: error,
+                })?;
+
+            return Err(UnitClientError::HttpResponseJsonBodyError {
+                status,
                 path: path.to_string(),
-            })
+                error: error.get("error").unwrap_or(&"Unknown error".into()).to_string(),
+                detail: error.get("detail").unwrap_or(&"".into()).to_string(),
+            });
+        }
+        serde_json::from_reader(reader).map_err(|error| UnitClientError::JsonError {
+            source: error,
+            path: path.to_string(),
         })
     }
 
     pub fn listeners_api(&self) -> Box<dyn ListenersApi + 'static> {
         new_openapi_client!(self, ListenersApiClient, ListenersApi)
     }
 
-    pub fn listeners(&self) -> Result<HashMap<String, ConfigListener>, Box<UnitClientError>> {
-        let list_listeners = self.listeners_api().get_listeners();
-        self.execute_openapi_future(list_listeners)
-    }
-
-    pub fn execute_openapi_future<F: Future<Output = Result<R, OpenAPIError>>, R: for<'de> serde::Deserialize<'de>>(
-        &self,
-        future: F,
-    ) -> Result<R, Box<UnitClientError>> {
-        self.rt.block_on(future).map_err(|error| {
-            let remapped_error = if let OpenAPIError::Hyper(hyper_error) = error {
-                UnitClientError::new(hyper_error, self.control_socket.to_string(), "".to_string())
+    pub async fn listeners(&self) -> Result<HashMap<String, ConfigListener>, Box<UnitClientError>> {
+        self.listeners_api().get_listeners().await.or_else(|err| {
+            if let OpenAPIError::Hyper(hyper_error) = err {
+                Err(Box::new(UnitClientError::new(
+                    hyper_error,
+                    self.control_socket.to_string(),
+                    "".to_string(),
```

Comment:
Do we want to add something like `"/listeners"` to the error, rather than an empty string?

Reply:
good idea! will add this.

Reply:
good idea! added.